### ISSUE TITLE
Allow Raft to force change configuration when there is no leader or quorum.

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#16148]
- RaftServer exposes a method to remove a set of brokers from the configuration
- Reconfiguration must complete even when there is no leader at the moment, and there is no quorum. However, a quorum among the remaining brokers is required.
- Reconfiguration must complete even when the brokers that are being removed is up and communicating with the remaining brokers.
- If the method times out, we can assume that it will be retried. This will simplify failure handling. 
- Reject reconfigure if the number of replicas < 1 after reconfiguration.
- Ensure there is no backward incompatible changes in raft messages.

Tasks
- [x] Define `Configuration` with flag "force". Extend SBE definition and serialization.
- [x] Reject any message from non-members if force reconfigure is in progress
- [x] Define `ForceReconfigure` request send between the followers and add handlers
- [x] Handle force configure in followers - the one which initiates as well as other followers.
- [x] Leader should come out of force configure immediately after transitioning
- [x] Add unit test for the behavior
- [x] Take the member type as input instead of assuming it to be ACTIVE

**Comment by deepthidevaki** [comment]
Here is the protocol implemented in POC [branch](https://github.com/camunda/zeebe/tree/dd-hw-failover-with-dynamic-config)

https://textik.com/#fbad72c9325dd380
```
                                                                                                                                                                               
                     Raft 0 (follower)                            Raft 1 (follower)                         Raft 2/3                                                           
                          |                                          |                                      (Members to be removed)                                            
 ForceReconfigure         |                                          |                                          |                                                              
 (newMembers=0,1)-------->|                                          |                                          |                                                              
                          |                                          |                                          |                                                              
        Configuration={   |                                          |                                          |                                                              
         newMembers=[0,1],|                                          |                                          |                                                              
         oldMembers=[]    |                                          |                                          |                                                              
         force=TRUE       |                                          |                                          |                                                              
        Commit new config |            Force Reconfigure(newConfig)  |                                          |                                                              
                          |----------------------------------------->| Commit newConfig                         |                                                              
                          |                       OK                 |                                          |                                                              
                          |<-----------------------------------------|                                          |                                                              
                          |                                          |        Poll/Vote/Append                  |                                                              
                          |                                          |<-----------------------------------------|                                                              
                 election |                  poll/vote               |----------------------------------------->|                                                              
                 timeout  ------------------------------------------>|     Reject because Force==TRUE           |                                                              
                          |                       OK                 |                                          |                                                              
                          |<-----------------------------------------|                                          |                                                              
             Become leader|                                          |                                          |                                                              
         Update InMemory  |                                          |                                          |                                                              
         Configuration={  |                       ConfigureRequest   |                                          |                                                              
          newMembers=[0,1]|----------------------------------------->|Update InMemory                           |                                                              
          force=FALSE     |                                          |Configuration= {                          |                                                              
         }                |                                          | newMembers=[0,1]                         |                                                              
                          |                                          | force=FALSE                              |                                                              
              Write       |                      AppendEntry         |}                                         |                                                              
              InitialEntry|----------------------------------------->|                                          |                                                              
                          |<-----------------------------------------|                                          |                                                              
                          |                                          |                                          |                                                              
         Commit new config|                      AppendEntry         |On commitIndex update                     |                                                              
                          |----------------------------------------->|Commit new config                         |                                                              
                          |                                          |                                          |                                                              
                          |                                          |       Poll/Vote                          |                                                              
                          |                                          |<-----------------------------------------|                                                              
                          |                                          |      Reject because log not uptodate     |                                                              
                          |                                          |----------------------------------------->|                                                              
                          |                                          |                                          |                                                              
                          |                                          |                                          |                                                              
                          |                                          |                                          |                                                              
                                                                     |                                          |                                                              
                                                                                                                                                                               
                                                                                                                                   
```




The main challenges are:
1. Configuration change is completed even when there is no leader
2. The brokers that are to be removed, should not block the progress of making configuration change.

During a normal operation, a configuration change is done via [Joint Consensus](https://github.com/camunda/zeebe/issues/13906) which is always initiated and coordinated by a leader. But here, we don't have a leader and we don't have quorum. To overcome this we introduce a new 
"Force Consensus" (for lack of a better term). You can pick any follower to start the force consensus. The follower enters the force consensus by writing a new Configuration with only the new members, and setting a flag `force=TRUE`. It immediately commits this configuration by writing it to a file. (Normally, this should have been replicated to a quorum and only after that it can be committed, but we are force overwriting the config). 
Then this follower sends the new config to all remaining members. All remaining members will also commit the new config with `force=TRUE`.

Now one of the follower hits the election timeout, and starts election. It goes through the normal election process. But now the quorom has changed because the new configuration contains only a subset of the members. When the follower becomes a leader, it comes out of the "force consensus" by writing a new ConfigurationEntry, replicating it to the new quorum and committing it following normal raft protocol. Once the new configuration is committed, we are back to normal raft protocol. 

When a member is in force consensus (i.e when force=TRUE), it will not accept any request from members that are not part of the config. (This is different from joint consensus. During normal operations, it is required that members accept request from any member to handle the scenario where the local config is outdated.). This behavior is required to ensure that the brokers to me removed do not interfere with the configuration change.

PS:- The above is what we implemented in POC. We still have to validate if this covers all edge cases.

**Comment by deepthidevaki** [comment]
One edge case to look into:
- What if the follower we chose to start force consensus has an outdated Configuration.

**Comment by deepthidevaki** [comment]
Here are some edge cases to help the discussions and writing tests.

1. **Raft 0 writes new force config the disk, but request to Raft 1 fails.**
2. **Raft 1 received force Reconfigure and write it to disk, but the response fails.**
3. **When Raft 0 receives the config - it is the leader with quorum (0,2 and 3)**
4. **When Raft 1 receives the config - it is the leader with quorum (1,2,3)**
5. **Can we force remove 2 if if the current members are 0, 1 and 2?**
6. **Can we force remove 1 and 2 if the current member are 0, 1 and 2?**
7. **After Raft 1 response OK for Force Reconfigure, Raft 0 encounters a data loss.**
8. **After Raft 1 response OK, Raft 1 encounters a data loss.**
9. **Can Raft 0 accepts and writes force config, but Raft 1 rejects it for some reason? Resulting in a never ending retry loop.** 
10. **Can Raft 2 or 3 gets a quorum after force config has been committed and new leader among 0 and 1 is elected? As a result overriding the new config.**
11. **What happens if we call force Reconfigure 0,1 and force Reconfigure 2,3 at the same time?**
12. **Raft 0 has an outdated config (may be because it was away for a long time). Can it result in an incorrect configuration overriding the latest config in Raft 1?** 

TODO: I will add how each of these are handled, or if it is not handled what is the expected behavior.

**Comment by deepthidevaki** [comment]
We identified one edge case in this PR https://github.com/camunda/zeebe/pull/16295

1. Member X receives a force configure request and accepts it
2. Member X got disconnected from the cluster
3. Other members of the cluster forms the quorum and committed new configuration
4. Later new force configure request was initiated
5. Member X which was disconnected all this time, was reconnected and received the new force configure request

In this case, we want Member X to accept the request. However, we also want to reject concurrent conflicting force configure requests. Distinguishing these two cases are tricky. We are yet to find a solution for this.

**Comment by deepthidevaki** [comment]
Regarding the detection of concurrent conflicting force configure request, I'm thinking of not handling it. Since it is a "force configure" it is not easy to correctly detect and resolve such conflicts. As we saw, sometimes the conflict is not an actual conflict but it is an out-dated follower which should accept the new force configuration. 

Since it is an inherently unsafe operation, it is the responsibility of the caller to ensure that there is no conflicting force configure. For example, in this case, we can probably prevent some cases in the `ClusterTopologyManager` before starting the operation. But at the end the user should ensure that they do not issue two concurrent force requests. 

@npepinpe  My proposal is to remove the "rejection" path when handling the force configure request. We can detect it and log for observability. But always accept and overwrite the local configuration even if it is conflicting. What do you think?

**Comment by npepinpe** [comment]
I think from past experience, we know that "it is the responsibility of the caller to ensure" will eventually happen (i.e. see multiple brokers with the same IDs :smile:). I think it's fine to accept that and delegate it to users, but we need to make it as easy to detect as possible, and decide on the simplest failure mode.

So what would happen if we always accept and overwrite the local configuration? Are we in a worst state? Do we need a "verification" step before allowing a cluster to come out of forced configuration? :thinking: e.g. if we detected conflicts, then instead of automatically getting out of the force state, we would require an extra request from the caller to confirm? Or are the consequences not bad enough to warrant this extra paranoia?

**Comment by deepthidevaki** [comment]
> So what would happen if we always accept and overwrite the local configuration? Are we in a worst state?

Good question 
Answer: No difference. We are screwed anyway. :grin: 

If we always accept and overwrite the local configuration, the second "force configure" wins. 
Let's say we force configure 0,1 and force configure 1,2 in parallel - 

1. 0,1 committed force configure
2. 1 received new force configure 1,2
3. 1 accepted new force configure 1,2
4. 0 tries to become leader => fails 
5. 1,2 elect new leader => forms the new group

Another case: force configure 0,1,2 and force configure 2,3,4 (You shouldn't be force configuring in this case, anyway)
1. 0,1,2 committed force configure
2. 2 received new force configure 2,3,4
3. 2 accepted new force configure 2,3,4
4. 0,1 forms quorum and elects the leader
5. 2,3,4 forms quorum and elects the leader => Split brain

If we do not accept the conflicting request, and reject it in case 1
1. 0,1 committed force configure
2. 1 received new force configure 1,2
3. 1 rejected new force configure 1,2
4. 0,1 forms quorum and comes out of force configuration
5. 1 receives new force configure 1,2 again (because it was retried)
6. Accepts and overwrite (because local is not force)
7. 1,2 forms the new members.

Similar in case 2.


Considering this feature is expected to be used in 2-region, the users are expected to call only one of the following:
- force configure 0,2 OR
- force configure 1,3 

If they call both, we cannot detect the conflict in raft anyway because they are disjoint sets of members.

**Comment by deepthidevaki** [comment]
Based on our [discussion](https://github.com/camunda/zeebe/issues/16145#issuecomment-1946255138) we can assume that there won't be any conflicting force configure request received by raft. So following my proposal, we can always accept force configure request. There are no cases where we want to reject it. For observability, we can still log if we suspect that a force configure request is conflicting with the local configuration.
